# Setup

In [1]:
import numpy as np
import pandas as pd
import geopandas as gpd
import pickle
import sys
from pathlib import Path

# Add the src directory to the path
sys.path.insert(0, '/Users/yew/code/neon-agbd/src')

from src.main import compute_site_biomass_full, ALL_SITES

In [2]:
site_id = 'DELA'

# Run the full workflow
output = compute_site_biomass_full(
	site_id=site_id,
	dp1_data_dir="./data/DP1.10098",
	agb_data_dir="./data/NEONForestAGB",
	plot_polygons_path="./data/plot_polygons/NEON_TOS_Plot_Polygons.geojson",
	apply_gap_filling=True,
	apply_dead_corrections=True,
	verbose=True
)

Processing site: DELA
  Loading DP1.10098 data...
  Loading NEONForestAGB data...
  Merging AGB estimates with apparent individual data...
  Loading plot area data...
  Categorizing individuals (tree vs small_woody)...
  Applying dead status corrections...
  Merging AGB estimates with apparent individual data...
  Loading plot area data...
  Categorizing individuals (tree vs small_woody)...
  Applying dead status corrections...
  Identifying unaccounted trees...
  Identifying unaccounted trees...
  Computing plot-level biomass...
  Computing plot-level biomass...
  Calculating growth metrics...
  Creating individual tree table...
  Calculating growth metrics...
  Creating individual tree table...
  Done! Computed biomass for 142 plot-year combinations.
  Found 507 unaccounted trees.
  Created individual tree table with 5283 records.
  Done! Computed biomass for 142 plot-year combinations.
  Found 507 unaccounted trees.
  Created individual tree table with 5283 records.


In [3]:
input_file = f'/Users/yew/code/neon-agbd/data/DP1.10098/{site_id}.pkl'
with open(input_file, 'rb') as f:
	input = pickle.load(f)
# dict_keys(['variables_10098', 'vst_apparentindividual', 'vst_mappingandtagging', 'vst_perplotperyear'])


In [18]:
dpids= ['DELA','LENO','TALL','BONA','DEJU','HEAL','SRER','SJER','SOAP',
              'TEAK','CPER','NIWO','RMNP','DSNY','OSBS','JERC','PUUM','KONZ',
              'UKFS','SERC','HARV','UNDE','BART','JORN','DCFS','NOGP','WOOD',
              'GUAN','LAJA','GRSM','ORNL','CLBJ','MOAB','ONAQ','BLAN','MLBS',
              'SCBI','ABBY','WREF','STEI','TREE','YELL']

out_list = []

for site_id in dpids:
    
	input_file = f'/Users/yew/code/neon-agbd/data/DP1.10098/{site_id}.pkl'
	with open(input_file, 'rb') as f:
		input = pickle.load(f)
	out_list.extend(list(np.unique(input['vst_apparentindividual']['plantStatus'])))
	
	# dict_keys(['variables_10098', 'vst_apparentindividual', 'vst_mappingandtagging', 'vst_perplotperyear'])


In [20]:
np.unique(out_list)

array(['', 'Dead, broken bole', 'Downed', 'Live', 'Live,  other damage',
       'Live, broken bole', 'Live, disease damaged',
       'Live, insect damaged', 'Live, physically damaged', 'Lost, burned',
       'Lost, fate unknown', 'Lost, herbivory', 'Lost, presumed dead',
       'Lost, tag damaged', 'No longer qualifies', 'Removed',
       'Standing dead'], dtype='<U24')

In [4]:
# dict_keys(['plot_biomass', 'unaccounted_trees', 'individual_trees', 'site_id', 'metadata'])
o_plots = output['plot_biomass']
o_plots

siteID    plotID  year  plotArea_m2  tree_AGBJenkins  tree_AGBChojnacky  \
0     DELA  DELA_001  2019         1600        66.244231          64.395718   
1     DELA  DELA_001  2022         1600        82.694043          80.172850   
2     DELA  DELA_002  2019         1600        27.295091          26.750160   
3     DELA  DELA_002  2022         1600        29.948876          29.826796   
4     DELA  DELA_003  2019         1600        77.464038          73.286685   
..     ...       ...   ...          ...              ...                ...   
137   DELA  DELA_056  2015         1600       116.586412          88.584163   
138   DELA  DELA_056  2016         1600       120.261858          92.121703   
139   DELA  DELA_056  2017         1600       126.945015          95.057212   
140   DELA  DELA_056  2021         1600       141.680965         105.217830   
141   DELA  DELA_059  2021         1600       102.869205          86.757612   

     tree_AGBAnnighofer  n_trees  small_woody_AGBJenkins  \
0                   0.0       22                0.949005   
1                   0.0       22                0.598066   
2                   0.0       18                2.137501   
3                   0.0       17                3.143433   
4                   0.0       31                0.996915   
..                  ...      ...                     ...   
137                 0.0       40                1.032120   
138                 0.0       42                1.013630   
139                 0.0       41                0.839620   
140                 0.0       33                0.520076   
141                 0.0       44                0.085751   

     small_woody_AGBChojnacky  small_woody_AGBAnnighofer  n_small_woody_total  \
0                    0.983348                   0.189293                   29   
1                    0.617966                   0.108910                   21   
2                    2.379093                        NaN                   31   
3                    3.223156                        NaN                   25   
4                    1.144914                   0.495193                   50   
..                        ...                        ...                  ...   
137                  2.431035                   0.154987                   63   
138                  2.517907                   0.145180                   61   
139                  2.152970                   0.201638                   62   
140                  0.994226                   0.143436                   46   
141                  0.121422                   0.151519                   27   

     n_small_woody_measured  n_unaccounted_trees  total_AGBJenkins  \
0                        18                    1         67.193235   
1                        16                    1         83.292109   
2                        17                    1         29.432592   
3                        11                    1         33.092309   
4                        11                    0         78.460953   
..                      ...                  ...               ...   
137                      36                    8        117.618532   
138                      34                    8        121.275487   
139                      35                    8        127.784634   
140                      16                    8        142.201041   
141                      17                    1        102.954956   

     total_AGBChojnacky  total_AGBAnnighofer    growth  growth_cumu  
0             65.379065             0.189293       NaN     5.366291  
1             80.790816             0.108910  5.366291     5.366291  
2             29.129254             0.000000       NaN     1.219906  
3             33.049952             0.000000  1.219906     1.219906  
4             74.431600             0.495193       NaN     0.234485  
..                  ...                  ...       ...          ...  
137           91.015198       

# Debug 

#### Looking into plot DELA_037, where there are 118 missing trees. Need to see what is up!

In [10]:
plotID = 'DELA_046'
tree_gfs = ['single bole tree','multi-bole tree','small tree']
tree_size = 10

In [11]:
aidf = input['vst_apparentindividual']
i_trees = aidf[aidf.plotID == plotID][aidf.growthForm.isin(tree_gfs)][aidf.stemDiameter >= tree_size]
i_trees = i_trees[['individualID','stemDiameter', 'plotID','subplotID', 'dataQF','eventID','growthForm','plantStatus','height',]]

/var/folders/xb/d4z3wkh548x206c5wbvfjk7h0000gn/T/ipykernel_80272/146869680.py:2: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.

/var/folders/xb/d4z3wkh548x206c5wbvfjk7h0000gn/T/ipykernel_80272/146869680.py:2: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.



In [12]:
o_plots[o_plots.plotID == plotID]

siteID    plotID  year  plotArea_m2  tree_AGBJenkins  tree_AGBChojnacky  \
100   DELA  DELA_046  2015         1600       166.720658         122.982931   
101   DELA  DELA_046  2016         1600       159.352204         118.245630   
102   DELA  DELA_046  2017         1600       162.243175         119.945073   
103   DELA  DELA_046  2021         1600        71.416695          48.942082   

     tree_AGBAnnighofer  n_trees  small_woody_AGBJenkins  \
100                 0.0       98                1.944954   
101                 0.0       98                2.132075   
102                 0.0       97                1.841896   
103                 0.0       51                1.483218   

     small_woody_AGBChojnacky  small_woody_AGBAnnighofer  n_small_woody_total  \
100                  1.894058                   0.614046                  163   
101                  1.946305                   0.637921                  167   
102                  1.589325                   0.657223                  168   
103                  1.226762                   0.676997                  159   

     n_small_woody_measured  n_unaccounted_trees  total_AGBJenkins  \
100                     112                    3        168.665612   
101                     112                    3        161.484279   
102                     111                    3        164.085071   
103                      98                    3         72.899914   

     total_AGBChojnacky  total_AGBAnnighofer     growth  growth_cumu  
100          124.876990             0.614046        NaN   -16.819256  
101          120.191935             0.637921  -7.181334   -16.819256  
102          121.534398             0.657223   2.600792   -16.819256  
103           50.168844             0.676997 -22.796289   -16.819256

In [67]:
aidf = input['vst_apparentindividual']

test = input['vst_apparentindividual']

test[test.individualID == 'NEON.PLA.D08.DELA.00373']

# Missing
# NEON.PLA.D08.DELA.03517

# Present
# NEON.PLA.D08.DELA.00373


# input['vst_apparentindividual']


uid  ...       release
213  506fe82a-b6e0-4125-96b1-99e1846b0b50  ...  RELEASE-2025

[1 rows x 43 columns]

In [ ]:
output['']

siteID  ...                                             reason
0     DELA  ...                 Never measured in survey campaigns
1     DELA  ...                 Never measured in survey campaigns
2     DELA  ...                 Never measured in survey campaigns
3     DELA  ...                 Never measured in survey campaigns
4     DELA  ...                 Never measured in survey campaigns
..     ...  ...                                                ...
502   DELA  ...  Has diameter measurements but no biomass estim...
503   DELA  ...  Has diameter measurements but no biomass estim...
504   DELA  ...  Has diameter measurements but no biomass estim...
505   DELA  ...  Has diameter measurements but no biomass estim...
506   DELA  ...  Has diameter measurements but no biomass estim...

[507 rows x 7 columns]

In [13]:
output['individual_trees']

siteID    plotID             individualID  year  AGBJenkins  \
0      DELA  DELA_053  NEON.PLA.D08.DELA.00373  2015    0.000000   
1      DELA  DELA_053  NEON.PLA.D08.DELA.00373  2016    0.000000   
2      DELA  DELA_053  NEON.PLA.D08.DELA.00373  2017    0.000000   
3      DELA  DELA_053  NEON.PLA.D08.DELA.00373  2021    0.000000   
4      DELA  DELA_046  NEON.PLA.D08.DELA.01166  2015    0.000000   
...     ...       ...                      ...   ...         ...   
5278   DELA  DELA_037  NEON.PLA.D08.DELA.13926  2019   28.778842   
5279   DELA  DELA_037  NEON.PLA.D08.DELA.13926  2020   28.778842   
5280   DELA  DELA_037  NEON.PLA.D08.DELA.13926  2021   28.778842   
5281   DELA  DELA_037  NEON.PLA.D08.DELA.13926  2022   28.778842   
5282   DELA  DELA_037  NEON.PLA.D08.DELA.13926  2023   28.778842   

      AGBChojnacky  AGBAnnighofer  growth_AGBJenkins  growth_cumu_AGBJenkins  \
0         0.000000            0.0                NaN                     0.0   
1         0.000000            0.0                0.0                     0.0   
2         0.000000            0.0                0.0                     0.0   
3         0.000000            0.0                0.0                     0.0   
4         0.000000            0.0                NaN                     0.0   
...            ...            ...                ...                     ...   
5278     40.257694            0.0                0.0                     0.0   
5279     40.257694            0.0                0.0                     0.0   
5280     40.257694            0.0                0.0                     0.0   
5281     40.257694            0.0                0.0                     0.0   
5282     40.257694            0.0                0.0                     0.0   

      growth_AGBChojnacky  ...  corrected_is_dead  gapFilling  \
0                     NaN  ...              False    ORIGINAL   
1                     0.0  ...              False      FILLED   
2                     0.0  ...              False      FILLED   
3                     0.0  ...              False      FILLED   
4                     NaN  ...              False      FILLED   
...                   ...  ...                ...         ...   
5278                  0.0  ...              False      FILLED   
5279                  0.0  ...              False      FILLED   
5280                  0.0  ...              False      FILLED   
5281                  0.0  ...              False    ORIGINAL   
5282                  0.0  ...              False    ORIGINAL   

                       scientificName  taxonID     genus      family  \
0                                 NaN      NaN       NaN         NaN   
1                                 NaN      NaN       NaN         NaN   
2                                 NaN      NaN       NaN         NaN   
3                                 NaN      NaN       NaN         NaN   
4     Fraxinus pennsylvanica Marshall     FRPE  Fraxinus    Oleaceae   
...                               ...      ...       ...         ...   
5278      Carpinus caroliniana Walter   CACA18  Carpinus  Betulaceae   
5279      Carpinus caroliniana Walter   CACA18  Carpinus  Betulaceae   
5280      Carpinus caroliniana Walter   CACA18  Carpinus  Betulaceae   
5281      Carpinus caroliniana Walter   CACA18  Carpinus  Betulaceae   
5282      Carpinus caroliniana Walter   CACA18  Carpinus  Betulaceae   

     taxonRank pointID stemDistance stemAzimuth  
0          NaN     NaN          NaN         NaN  
1          NaN     NaN          NaN         NaN  
2          NaN     NaN          NaN         NaN  
3          NaN     NaN          NaN         NaN  
4      species      39          3.7        87.5  
...        ...     ...          ...         ...  
5278   species      43         11.7       203.8  
5279   species      43         11.7       203.8  
5280   species      43         11.7       203.8  
5281   species      43         11.7       203.8  
5282   species      43         

In [8]:
o_utrees = output['unaccounted_trees']
o_utrees

siteID    plotID              individualID  \
0     DELA  DELA_053   NEON.PLA.D08.DELA.04229   
1     DELA  DELA_037   NEON.PLA.D08.DELA.04965   
2     DELA  DELA_007   NEON.PLA.D08.DELA.09232   
3     DELA  DELA_039   NEON.PLA.D08.DELA.11111   
4     DELA  DELA_002  NEON.PLA.D08.DELA.14329Z   
..     ...       ...                       ...   
502   DELA  DELA_041   NEON.PLA.D08.DELA.09719   
503   DELA  DELA_038   NEON.PLA.D08.DELA.10183   
504   DELA  DELA_038   NEON.PLA.D08.DELA.10196   
505   DELA  DELA_037   NEON.PLA.D08.DELA.13903   
506   DELA  DELA_037   NEON.PLA.D08.DELA.04941   

                        scientificName taxonID        status  \
0    Berchemia scandens (Hill) K. Koch    BESC    UNMEASURED   
1      Fraxinus pennsylvanica Marshall    FRPE    UNMEASURED   
2              Diospyros virginiana L.   DIVI5    UNMEASURED   
3                     Quercus nigra L.    QUNI    UNMEASURED   
4      Fraxinus pennsylvanica Marshall    FRPE    UNMEASURED   
..                                 ...     ...           ...   
502                     Acer rubrum L.    ACRU  NO_ALLOMETRY   
503                   Quercus nigra L.    QUNI  NO_ALLOMETRY   
504                   Quercus nigra L.    QUNI  NO_ALLOMETRY   
505                   Quercus nigra L.    QUNI  NO_ALLOMETRY   
506                     Acer rubrum L.    ACRU  NO_ALLOMETRY   

                                                reason  
0                   Never measured in survey campaigns  
1                   Never measured in survey campaigns  
2                   Never measured in survey campaigns  
3                   Never measured in survey campaigns  
4                   Never measured in survey campaigns  
..                                                 ...  
502  Has diameter measurements but no biomass estim...  
503  Has diameter measurements but no biomass estim...  
504  Has diameter measurements but no biomass estim...  
505  Has diameter measurements but no biomass estim...  
506  Has diameter measurements but no biomass estim...  

[507 rows x 7 columns]

In [69]:
o_plots

siteID    plotID  year  ...  total_AGBAnnighofer    growth  growth_cumu
0     DELA  DELA_001  2019  ...             0.169711       NaN     5.671804
1     DELA  DELA_001  2022  ...             0.103724  5.671804     5.671804
2     DELA  DELA_002  2019  ...             0.000000       NaN     1.219906
3     DELA  DELA_002  2022  ...             0.000000  1.219906     1.219906
4     DELA  DELA_003  2019  ...             0.452564       NaN     0.289328
..     ...       ...   ...  ...                  ...       ...          ...
137   DELA  DELA_056  2015  ...             0.118745       NaN     4.254736
138   DELA  DELA_056  2016  ...             0.104720  3.605616     4.254736
139   DELA  DELA_056  2017  ...             0.152854  6.936499     4.254736
140   DELA  DELA_056  2021  ...             0.140317  3.726359     4.254736
141   DELA  DELA_059  2021  ...             0.151519       NaN          NaN

[142 rows x 19 columns]

## Interactive version with Plotly (hover to see plotID)

In [9]:
import plotly.express as px

# Create interactive plot with Plotly
fig = px.line(o_plots, 
              x='year', 
              y='total_AGBJenkins',
              color='plotID',
              markers=True,
              labels={'total_AGBJenkins': 'AGB Jenkins (T/ha)',
                      'year': 'Year',
                      'plotID': 'Plot ID'},
              title='Above-Ground Biomass Over Time by Plot (Interactive)',
              hover_data={'plotID': True, 'year': True, 'total_AGBJenkins': ':.2f'})

# Customize layout
fig.update_traces(marker=dict(size=8), line=dict(width=2))
fig.update_layout(
    height=600,
    width=1000,
    hovermode='closest',
    legend=dict(
        orientation="h",
        yanchor="bottom",
        y=-0.3,
        xanchor="center",
        x=0.5,
        title="Plot ID"
    ),
    xaxis=dict(title_font=dict(size=14, family='Arial, bold')),
    yaxis=dict(title_font=dict(size=14, family='Arial, bold')),
    title_font=dict(size=16, family='Arial, bold')
)

fig.show()

print(f"\nInteractive plot created! Hover over points to see details.")
print(f"You can also: zoom, pan, select regions, and toggle plot visibility by clicking legend items.")


Interactive plot created! Hover over points to see details.
You can also: zoom, pan, select regions, and toggle plot visibility by clicking legend items.
